homeassistant/components/rest/__init__.py

```diff
@@ -1,4 +1,170 @@
 """The rest component."""
 
-DOMAIN = "rest"
+import asyncio
+import logging
+
+import httpx
+import voluptuous as vol
+
+from homeassistant.components.binary_sensor import DOMAIN as BINARY_SENSOR_DOMAIN
+from homeassistant.components.sensor import DOMAIN as SENSOR_DOMAIN
+from homeassistant.const import (
+    CONF_AUTHENTICATION,
+    CONF_HEADERS,
+    CONF_METHOD,
+    CONF_PARAMS,
+    CONF_PASSWORD,
+    CONF_PAYLOAD,
+    CONF_RESOURCE,
+    CONF_RESOURCE_TEMPLATE,
+    CONF_SCAN_INTERVAL,
+    CONF_TIMEOUT,
+    CONF_USERNAME,
+    CONF_VERIFY_SSL,
+    HTTP_DIGEST_AUTHENTICATION,
+    SERVICE_RELOAD,
+)
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.helpers import discovery
+from homeassistant.helpers.entity_component import (
+    DEFAULT_SCAN_INTERVAL,
+    EntityComponent,
+)
+from homeassistant.helpers.reload import async_reload_integration_platforms
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import COORDINATOR, DOMAIN, PLATFORM_IDX, REST, REST_IDX
+from .data import RestData
+from .schema import CONFIG_SCHEMA  # noqa:F401 pylint: disable=unused-import
+
+_LOGGER = logging.getLogger(__name__)
+
 PLATFORMS = ["binary_sensor", "notify", "sensor", "switch"]
+COORDINATOR_AWARE_PLATFORMS = [SENSOR_DOMAIN, BINARY_SENSOR_DOMAIN]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the rest platforms."""
+    component = EntityComponent(_LOGGER, DOMAIN, hass)
+    _async_setup_shared_data(hass)
+
+    async def reload_service_handler(service):
+        """Remove all user-defined groups and load new ones from config."""
+        conf = await component.async_prepare_reload()
+        if conf is None:
+            return
+        await async_reload_integration_platforms(hass, DOMAIN, PLATFORMS)
+        _async_setup_shared_data(hass)
+        await _async_process_config(hass, conf)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_RELOAD, reload_service_handler, schema=vol.Schema({})
+    )
+
+    return await _async_process_config(hass, config)
+
+
+@callback
+def _async_setup_shared_data(hass: HomeAssistant):
+    """Create shared data for platform config and rest coordinators."""
+    hass.data[DOMAIN] = {platform: {} for platform in COORDINATOR_AWARE_PLATFORMS}
+
+
+async def _async_process_config(hass, config) -> bool:
+    """Process rest configuration."""
+    if DOMAIN not in config:
+        return True
+
+    load_tasks = []
+    for rest_idx, conf in enumerate(config[DOMAIN]):
+        scan_interval = conf.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL)
+        resource_template = conf.get(CONF_RESOURCE_TEMPLATE)
+        rest = create_rest_data_from_config(hass, conf)
+        coordinator = _wrap_rest_in_coordinator(
+            hass, rest, resource_template, scan_interval
+        )
+        await coordinator.async_refresh()
```

Comment:
`await` inside a for-loop means that we will fetch all Rest endpoints in sequence instead of parallel.

Reply:
🤦 Almost got it right. haha. At least I remembered to gather this time :)

Reply:
fixed